File: ios/AtlasCore/Sources/AtlasCore/Spaces/SpacePolicy.swift
import Foundation

/// Space policy — the shared access substrate every vault sits on. Byte-for-byte parity with
/// `backend/atlas/spaces/space_policy.py` (Python is reference-of-record).
///
/// One mechanism — MEMBERSHIP + ROLE under POLICY — expresses sharing, business officer-quorum
/// governance, and agent capability-leashes alike. Every access change is a QUORUM-authorized entry
/// in an APPEND-ONLY, hash-chained log (audit trail + sync source of truth). `head()` is the
/// anchorable digest (signed hash-chain by default; publish it to the beacon for third-party-
/// verifiable immutability — only the head leaves, never content). Revocation is forward-only + logged.
public enum SpaceGovernance {

    static let changeLabel = Data("atlas/space-policy/change/v1".utf8)
    static let entryLabel = Data("atlas/space-policy/entry/v1".utf8)
    static let genesisLabel = Data("atlas/space-policy/genesis/v1".utf8)

    public enum PolicyError: Error, Equatable { case unknownOp, notAuthorized, stale, badChange }

    /// Ordered capability levels — a higher role includes every lower capability.
    public enum Role: Int, Comparable, Sendable {
        case reader = 1, contributor = 2, governor = 3, breakGlass = 4
        public static func < (l: Role, r: Role) -> Bool { l.rawValue < r.rawValue }
    }

    static func lp(_ d: Data) -> Data {
        var n = UInt32(d.count).bigEndian
        return withUnsafeBytes(of: &n) { Data($0) } + d
    }
    static func u16(_ v: Int) -> Data { var n = UInt16(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }
    static func u32(_ v: Int) -> Data { var n = UInt32(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }
    static func u64(_ v: Int) -> Data { var n = UInt64(v).bigEndian; return withUnsafeBytes(of: &n) { Data($0) } }

    public struct Change: Equatable {
        public let op: String                 // "grant" | "revoke" | "set_quorum"
        public let target: Data               // encoded HybridSign.PublicKey (or empty)
        public let role: Role?
        public let quorum: Int?
        public let epoch: Int
        public let afterHead: Data

        public func body() -> Data {
            changeLabel + lp(Data(op.utf8)) + lp(target)
                + u16(role?.rawValue ?? 0) + u32(quorum ?? 0) + u64(epoch) + lp(afterHead)
        }
    }

    public struct LogEntry: Equatable {
        public let seq: Int
        public let prevHash: Data
        public let changeBody: Data
        public let approvers: [Data]          // encoded governor publics who authorized
        public let entryHash: Data

        public static func computeHash(prevHash: Data, changeBody: Data, approvers: [Data]) -> Data {
            // H concatenates its chunks; all fields are length-prefixed, so one buffer == variadic H.
            var buf = entryLabel + lp(prevHash) + lp(changeBody) + u32(approvers.count)
            for a in approvers { buf += lp(a) }
            return Primitives.H(buf)
        }
    }

    /// The living access set + policy for one space, with its append-only authorization log.
    public final class SpacePolicy {
        public let spaceID: Data
        public private(set) var members: [Data: Role]   // encoded public -> Role
        public private(set) var quorum: Int
        public private(set) var log: [LogEntry]

        private init(spaceID: Data, members: [Data: Role], quorum: Int, log: [LogEntry]) {
            self.spaceID = spaceID; self.members = members; self.quorum = quorum; self.log = log
        }

        public static func genesis(spaceID: Data, creator: HybridSign.PublicKey, epoch: Int = 0) -> SpacePolicy {
            let enc = creator.encode()
            let gbody = Primitives.H(genesisLabel + lp(spaceID) + lp(enc) + u64(epoch))
            let entry = LogEntry(seq: 0, prevHash: Data(), changeBody: gbody, approvers: [enc],
                                 entryHash: LogEntry.computeHash(prevHash: Data(), changeBody: gbody, approvers: [enc]))
            return SpacePolicy(spaceID: spaceID, members: [enc: .governor], quorum: 1, log: [entry])
        }

        public func head() -> Data { log.last?.entryHash ?? Data() }
        public func roleOf(_ m: HybridSign.PublicKey) -> Role? { members[m.encode()] }
        public func can(_ m: HybridSign.PublicKey, _ capability: Role) -> Bool {
            guard let r = members[m.encode()] else { return false }
            return r >= capability
        }
        public func governors() -> Set<Data> { Set(members.filter { $0.value >= .governor }.keys) }
        public func membersAt(_ minimum: Role) -> [Data] { members.filter { $0.value >= minimum }.map { $0.key } }

        public func propose(_ op: String, target: HybridSign.PublicKey? = nil, role: Role? = nil,
                            quorum: Int? = nil, epoch: Int) throws -> Change {
            guard ["grant", "revoke", "set_quorum"].contains(op) else { throw PolicyError.unknownOp }
            return Change(op: op, target: target?.encode() ?? Data(), role: role, quorum: quorum,
                          epoch: epoch, afterHead: head())
        }

        @discardableResult
        public func authorize(_ change: Change,
                              approvals: [(HybridSign.PublicKey, Data)]) throws -> LogEntry {
            guard change.afterHead == head() else { throw PolicyError.stale }
            let body = change.body()
            let gov = governors()
            var seen: Set<Data> = []
            for (pub, sig) in approvals {
                let enc = pub.encode()
                if gov.contains(enc), !seen.contains(enc), HybridSign.verify(pub, body, sig) { seen.insert(enc) }
            }
            guard seen.count >= quorum else { throw PolicyError.notAuthorized }

            switch change.op {
            case "grant":
                guard let role = change.role else { throw PolicyError.badChange }
                members[change.target] = role
            case "revoke":
                members[change.target] = nil                 // forward-only; history stays in the log
            case "set_quorum":
                guard let q = change.quorum, q >= 1, q <= governors().count else { throw PolicyError.badChange }
                quorum = q
            default: throw PolicyError.unknownOp
            }

            let approvers = seen.sorted { $0.lexicographicallyPrecedes($1) }
            let entry = LogEntry(seq: log.count, prevHash: head(), changeBody: body, approvers: approvers,
                                 entryHash: LogEntry.computeHash(prevHash: head(), changeBody: body, approvers: approvers))
            log.append(entry)
            return entry
        }

        public func verifyLog() -> Bool {
            var prev = Data()
            for (i, e) in log.enumerated() {
                if e.seq != i || e.prevHash != prev { return false }
                if e.entryHash != LogEntry.computeHash(prevHash: e.prevHash, changeBody: e.changeBody, approvers: e.approvers) { return false }
                prev = e.entryHash
            }
            return true
        }
    }
}

File: backend/atlas/spaces/space_policy.py
"""Space policy — the shared access substrate every vault sits on. A "folder" is a space; its
security is a POLICY on that space. Records assembly, the sensitive-records tier, anonymous-eligible
polls, and the vault all compose this one layer (see the entity/vault/Real-ID model).

One mechanism — MEMBERSHIP + ROLE, under POLICY — expresses everything discussed:
  * sharing a doc with a person   = add their handle as READER/CONTRIBUTOR
  * a business's officers         = GOVERNORs; changing the access set needs a governor QUORUM
  * an agent's leash              = a capability-scoped role; revoking = removing it
  * "independent read, joint governance" = READ is per-member; access CHANGES need M-of-N governors

Every access change is a QUORUM-authorized entry in an APPEND-ONLY, hash-chained log — which is both
the audit trail (retention ≠ readability; a coercer can't silently rewrite who-had-access) and the
sync source of truth (replay the log → both ends converge). `head()` is the anchorable digest:
signed-hash-chain by default; publish it to the drand beacon (or a public ledger) for third-party-
verifiable immutability — only the 32-byte head ever leaves, never content.

Revocation is FORWARD-ONLY (remove the role; rotate keys out of band) and always logged — honest,
the same as every crypto-sharing system. Not new crypto: HybridSig quorum over domain-separated,
length-prefixed bodies + a hash chain.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum
from typing import Dict, List, Sequence, Tuple

from ..crypto.primitives import H
from ..crypto.sign import HybridSigPublic, verify

_CHANGE = b"atlas/space-policy/change/v1"
_ENTRY = b"atlas/space-policy/entry/v1"
_GENESIS = b"atlas/space-policy/genesis/v1"


class PolicyError(Exception):
    pass


class NotAuthorized(PolicyError):
    """A change lacked a governor quorum, or an approver wasn't a current governor."""


class StaleChange(PolicyError):
    """The change was authorized against a different log head — someone else moved the chain."""


class Role(IntEnum):
    """Ordered capability levels — a higher role includes every lower capability."""
    READER = 1        # read space content
    CONTRIBUTOR = 2   # + add/update content
    GOVERNOR = 3      # + change the access set / policy (requires quorum)
    BREAK_GLASS = 4   # + emergency reveal of sealed material (always logged)


def _lp(b: bytes) -> bytes:
    return len(b).to_bytes(4, "big") + b


@dataclass(frozen=True)
class Change:
    """A proposed mutation of the access set / policy. `target` is the affected member's public key
    (encoded); `role` is the new role for a grant (None for a revoke). `set_quorum` changes the
    governor threshold. Bound to `after_head` so it can only apply to the exact chain state it was
    approved against (no replay onto a moved chain)."""
    op: str                       # "grant" | "revoke" | "set_quorum"
    target: bytes                 # encoded HybridSigPublic (or b"" for set_quorum)
    role: int | None              # Role for grant, else None
    quorum: int | None            # new quorum for set_quorum, else None
    epoch: int
    after_head: bytes             # the log head this change extends

    def body(self) -> bytes:
        return b"".join([_CHANGE, _lp(self.op.encode()), _lp(self.target),
                         (int(self.role) if self.role is not None else 0).to_bytes(2, "big"),
                         (self.quorum if self.quorum is not None else 0).to_bytes(4, "big"),
                         self.epoch.to_bytes(8, "big"), _lp(self.after_head)])


@dataclass(frozen=True)
class LogEntry:
    seq: int
    prev_hash: bytes
    change_body: bytes
    approvers: Tuple[bytes, ...]   # encoded governor publics who authorized (genesis: the creator)
    entry_hash: bytes

    @staticmethod
    def compute_hash(prev_hash: bytes, change_body: bytes, approvers: Sequence[bytes]) -> bytes:
        parts = [_ENTRY, _lp(prev_hash), _lp(change_body), len(approvers).to_bytes(4, "big")]
        parts.extend(_lp(a) for a in approvers)
        return H(*parts)


@dataclass
class SpacePolicy:
    """The living access set + policy for one space, with its append-only authorization log."""
    space_id: bytes
    members: Dict[bytes, Role] = field(default_factory=dict)   # encoded public -> Role
    quorum: int = 1                                            # governor approvals to mutate access
    log: List[LogEntry] = field(default_factory=list)

    # ---- construction ------------------------------------------------------
    @classmethod
    def genesis(cls, space_id: bytes, creator: HybridSigPublic, *, epoch: int = 0) -> "SpacePolicy":
        """Create a space. The creator (an enrolled individual) is the first GOVERNOR, quorum=1.
        The genesis entry roots the hash chain; there is no prior head to approve against."""
        p = cls(space_id=space_id, members={creator.encode(): Role.GOVERNOR}, quorum=1)
        gbody = H(_GENESIS, _lp(space_id), _lp(creator.encode()), epoch.to_bytes(8, "big"))
        entry = LogEntry(seq=0, prev_hash=b"", change_body=gbody,
                         approvers=(creator.encode(),),
                         entry_hash=LogEntry.compute_hash(b"", gbody, (creator.encode(),)))
        p.log.append(entry)
        return p

    # ---- reads -------------------------------------------------------------
    def head(self) -> bytes:
        return self.log[-1].entry_hash if self.log else b""

    def role_of(self, member: HybridSigPublic) -> Role | None:
        return self.members.get(member.encode())

    def can(self, member: HybridSigPublic, capability: Role) -> bool:
        """Read is per-member and independent; higher capabilities need the matching role."""
        r = self.members.get(member.encode())
        return r is not None and r >= capability

    def governors(self) -> set[bytes]:
        return {h for h, r in self.members.items() if r >= Role.GOVERNOR}

    def members_at(self, minimum: Role) -> list[bytes]:
        """Encoded publics of members at or above `minimum` — the basis for a poll's eligible set."""
        return [h for h, r in self.members.items() if r >= minimum]

    # ---- proposals + authorization ----------------------------------------
    def propose(self, op: str, *, target: HybridSigPublic | None = None,
                role: Role | None = None, quorum: int | None = None, epoch: int) -> Change:
        if op not in ("grant", "revoke", "set_quorum"):
            raise PolicyError(f"unknown op {op!r}")
        return Change(op=op, target=target.encode() if target else b"",
                      role=int(role) if role is not None else None,
                      quorum=quorum, epoch=epoch, after_head=self.head())

    def authorize(self, change: Change,
                  approvals: Sequence[Tuple[HybridSigPublic, bytes]]) -> LogEntry:
        """Apply `change` iff ≥ quorum DISTINCT current governors signed it over the current head.
        Appends a hash-chained, signed log entry and mutates the access set. Idempotent replay is
        rejected via `after_head` (a change approved against an old head is stale)."""
        if change.after_head != self.head():
            raise StaleChange("change was approved against a different log head")

        body = change.body()
        gov = self.governors()
        seen: set[bytes] = set()
        for pub, sig in approvals:
            enc = pub.encode()
            if enc in gov and enc not in seen and verify(pub, body, sig):
                seen.add(enc)
        if len(seen) < self.quorum:
            raise NotAuthorized(f"need {self.quorum} governor approvals, got {len(seen)}")

        # apply
        if change.op == "grant":
            if change.role is None:
                raise PolicyError("grant needs a role")
            self.members[change.target] = Role(change.role)
        elif change.op == "revoke":
            self.members.pop(change.target, None)          # forward-only; history stays in the log
        elif change.op == "set_quorum":
            if change.quorum is None or change.quorum < 1:
                raise PolicyError("set_quorum needs quorum >= 1")
            if change.quorum > len(self.governors()):
                raise PolicyError("quorum cannot exceed the number of governors")
            self.quorum = change.quorum

        approvers = tuple(sorted(seen))
        entry = LogEntry(seq=len(self.log), prev_hash=self.head(), change_body=body,
                         approvers=approvers,
                         entry_hash=LogEntry.compute_hash(self.head(), body, approvers))
        self.log.append(entry)
        return entry

    # ---- audit -------------------------------------------------------------
    def verify_log(self) -> bool:
        """Recompute the hash chain end-to-end — any silent edit to a past entry breaks it."""
        prev = b""
        for i, e in enumerate(self.log):
            if e.seq != i or e.prev_hash != prev:
                return False
            if e.entry_hash != LogEntry.compute_hash(e.prev_hash, e.change_body, e.approvers):
                return False
            prev = e.entry_hash
        return True
